xonsh/environ.py

```diff
@@ -2154,6 +2176,81 @@ def deregister(self, name):
         self._vars.pop(name)
 
 
+class InternalEnvironDict(ChainMap):
+    """A dictionary which supports thread-local overrides.
+    There are two reasons we can't use ChainMap directly:
+    1. To use thread-local storage, we need to access the local().__dict__ directly each time to get the correct dict.
+    2. We want to set items to the local storage only if they we're explicitly set there.
+    """
+
+    def __init__(self):
+        self._global = {}
+        self._thread_local = threading.local()
+        super().__init__()
+
+    @property
+    def _local(self):
+        # As per local's the documentation, accessing its `__dict__` works fine (but must be done inside the thread).
+        return self._thread_local.__dict__
+
+    @property  # type: ignore
+    def maps(self):
+        # The 'maps' array needs to contain the thread-local dictionary every time we use it.
+        # We prefer getting from the local scope if possible.
+        return [self._local, self._global]
+
+    @maps.setter
+    def maps(self, _v):
+        # This is here for ChainMap.__init__.
+        pass
+
+    def __setitem__(self, key, value):
+        # If the value is overridden locally, set it locally.
+        local = self._local
+        if key in local:
+            local[key] = value
+        else:
+            self._global[key] = value
+
+    def __delitem__(self, key):
+        # If the value is overridden locally, delete it locally.
+        try:
+            del self._local[key]
+        except KeyError:
+            del self._global[key]
```

Comment:
Is there a guarantee here that if the `key` isn't in the locals that it will exist in the globals?  Or should there be an additional try-catch?

Reply:
There's no guarantee -- if it doesn't exist locally and globally a `KeyError` should be thrown